BedrockServer.cpp

```diff
@@ -2106,52 +2102,87 @@ void BedrockServer::_finishPeerCommand(unique_ptr<BedrockCommand>& command) {
+                // Otherwise create the object for this new socket.
+                SDEBUG("Accepting socket from '" << addr << "' on port '" << port->host << "'");
+                Socket socket(s, Socket::CONNECTED);
+                socket.addr = addr;
 
-            // If we got an error or no socket, done accepting for now.
-            if (s <= 0) {
-                break;
-            }
+                // If it came from a plugin, record that.
+                auto plugin = _portPluginMap.find(port);
+                if (plugin != _portPluginMap.end()) {
+                    socket.data = plugin->second;
+                }
 
-            // Otherwise create the object for this new socket.
-            SDEBUG("Accepting socket from '" << addr << "' on port '" << port->host << "'");
-            Socket socket(s, Socket::CONNECTED);
-            socket.addr = addr;
+                // And start up this socket's thread.
+                _outstandingSocketThreads++;
+                thread t;
+                bool threadStarted = false;
+                while (!threadStarted) {
+                    try {
+                        t = thread(&BedrockServer::handleSocket, this, move(socket), port == _controlPort, port == _commandPortPublic, port == _commandPortPrivate);
+                        threadStarted = true;
+                    } catch (const system_error& e) {
+                        // We don't care about this lock here from a performance perspective, it only happens when we
+                        // are unable to do any work anyway (i.e., we can't start threads).
+                        lock_guard<mutex> lock(_newSocketThreadBlockedMutex);
+                        if (!_newSocketThreadsBlocked) {
+                            // Block any new socket threads and warn.
+                            _newSocketThreadsBlocked = true;
+                            SWARN("Caught system_error in thread constructor (with " << _outstandingSocketThreads
+                                  << " threads): " << e.code() << ", message: " << e.what() << ", blocking new socket threads.");
+                            blockCommandPort("NOT_ENOUGH_THREADS");
+                        }
 
-            // If it came from a plugin, record that.
-            auto plugin = _portPluginMap.find(port);
-            if (plugin != _portPluginMap.end()) {
-                socket.data = plugin->second;
+                        // We just loop until socket threads are unblocked.
+                        SINFO("Waiting 1 more second for socket threads to be available.");
+                        sleep(1);
+                    }
+                }
+                try {
+                    t.detach();
+                } catch (const system_error& e) {
+                    SALERT("Caught system_error in thread detach: " << e.code() << ", message: " << e.what());
+                    throw;
```

Comment:
NAB if the above block throws, wouldn't `t.detach()` also throw? Meaning we'll close the command port for `NOT_ENOUGH_THREADS` and then immediately crash?

Reply:
The above will have succeeded if we got here. It's wrapped in a while loop.